internal/primitives/state-machine/overlayedchanges/ext.go

```diff
@@ -70,3 +82,713 @@ func compactLen(val uint) int {
+		ext_id = %s
+		child_info = %s
+		storage_root = %s
+		cached = %v`,
+		common.BytesToHex(leID(e.Id)),
+		common.BytesToHex(childInfo.StorageKey()),
+		root.String(),
+		cached,
+	)
+
+	return scale.MustMarshal(root)
+}
+
+// Index specified transaction slice and store it.
+func (e Ext[H, Hasher, B]) StorageIndexTransaction(index uint32, hash []byte, size uint32) {
+	logger.Tracef(
+		`target = state 
+		method = RenewTransactionIndex
+		ext_id = %s
+		index = %d
+		tx_hash = %s`,
+		common.BytesToHex(leID(e.Id)),
+		index,
+		common.BytesToHex(hash),
+	)
+
+	e.overlay.AddTransactionIndex(IndexOperationRenew{Extrinsic: index, Hash: hash})
+}
+
+// Start a new nested transaction.
+//
+// This allows to either commit or roll back all changes made after this call to the
+// top changes or the default child changes. For every transaction there cam be a
+// matching call to either [Ext.StorageRollbackTransaction] or [Ext.StorageCommitTransaction].
+// Any transactions that are still open after returning from runtime are committed
+// automatically.
+//
+// Changes made without any open transaction are committed immediately.
+func (e Ext[H, Hasher, B]) StorageStartTransaction() {
+	e.overlay.StartTransaction()
+}
+
+// Rollback the last transaction started by [Ext.StorageStartTransaction].
+// Any changes made during that storage transaction are discarded. Returns an error when
+// no transaction is open that can be closed.
+func (e Ext[H, Hasher, B]) StorageRollbackTransaction() error {
+	_ = e.overlay.RollbackTransaction()
+	return nil
+}
+
+// Commit the last transaction started by [Ext.StorageStartTransaction].
+// Any changes made during that storage transaction are committed. Returns an error when
+// no transaction is open that can be closed.
+func (e Ext[H, Hasher, B]) StorageCommitTransaction() error {
+	_ = e.overlay.CommitTransaction()
+	return nil
+}
+
+func (e Ext[H, Hasher, B]) limitRemoveFromBackend(
+	childInfo storage.ChildInfo,
+	prefix []byte,
+	maybeLimit *uint32,
+	startAt []byte,
+) ([]byte, uint32, uint32) {
+	iter, err := e.backend.Keys(statemachine.IterArgs{
+		ChildInfo: childInfo,
+		Prefix:    prefix,
+		StartAt:   startAt,
+	})
+
+	if err != nil {
+		logger.Debugf("error while iterating the storage: %w", err)
+	}
+
+	deleteCount := uint32(0)
+	loopCount := uint32(0)
+	var maybeNextKey []byte
+
+	for key, err := range iter.All() {
+		if err != nil {
+			logger.Debugf("error while iterating the storage: %w", err)
+			break
+		}
+
+		if maybeLimit != nil && *maybeLimit == loopCount {
+			maybeNextKey = key
+			break
+		}
+
+		var has bool
+		var overlay []byte
+
+		if childInfo != nil {
+			overlay, has = e.overlay.ChildStorage(childInfo, key)
+		} else {
+			overlay, has = e.overlay.Storage(key)
+		}
+
+		if has {
+			// not pending deletion from the backend - delete it.
+			if overlay != nil {
+				e.overlay.SetChildStorage(childInfo, key, nil)
+			} else {
+				e.overlay.SetStorage(key, nil)
+			}
+
+			deleteCount += 1
```

Comment:
we have saturating operations now via the `saturating` package I added ([link](https://github.com/ChainSafe/gossamer/blob/31b2325e44f8376e66d75ac82878c6bd51596fd8/internal/saturating/saturating.go#L26)).